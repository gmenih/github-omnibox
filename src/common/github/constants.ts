export const DEFAULT_FIRST_LOGINS: number = 100;
export const DEFAULT_FIRST_REPOS: number = 5;
export const DEFAULT_SCOPES: string[] = ['repo', 'read:org'];
export const GITHUB_API: string = 'https://api.github.com/graphql';
export const GITHUB_OAUTH_URL: string = 'https://github.com/login/oauth/authorize';
export const GITHUB_TOKEN_URL: string = 'https://github.com/login/oauth/access_token';
export const REDIRECT_URL: string = process.env.REDIRECT_URL || 'https://github.com/gmenih341/github-omnibox';
export const CLIENT_ID: string = process.env.CLIENT_ID || '';
export const CLIENT_SECRET: string = process.env.CLIENT_SECRET || '';
