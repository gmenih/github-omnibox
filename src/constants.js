export const GITHUB_API = 'https://api.github.com/graphql';
export const GITHUB_OAUTH_URL = 'https://github.com/login/oauth/authorize';
export const GITHUB_TOKEN_URL = 'https://github.com/login/oauth/access_token';
export const REDIRECT_URL = 'https://github.com/gmenih341/github-omnibox';
export const SEARCH_DEBOUNCE = 200;
export const OPTION_STRINGS = {
    GITHUB_TOKEN: '__github.token',
    GITHUB_LOGINS: '__github.logins',
    SEARCH_GLOBAL: '__github.search_global',
    SEARCH_DESC: '__github.search_desc',
    SEARCH_FORKED: '__github.search_forked',

    USE_HISTORY: '__extension.use_history',
    HISTORY: '__extension.search_history',
    UPDATE_LOGINS: '__extension.update_logins',
    OPTIONS_SHOWN: '__extension.options_shown_once',
    TOKEN_TYPE: '__extension.token_type',
};
