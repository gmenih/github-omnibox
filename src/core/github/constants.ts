export const DEFAULT_FIRST_LOGINS = 100;
export const DEFAULT_FIRST_REPOS = 5;
export const DEFAULT_SCOPES: string[] = ['repo:status', 'read:org'];
export const GITHUB_API = 'https://api.github.com/graphql';
export const GITHUB_OAUTH_URL = 'https://github.com/login/oauth/authorize';
export const GITHUB_TOKEN_URL = 'https://github.com/login/oauth/access_token';
export const REDIRECT_URL = 'https://github.com/gmenih341/github-omnibox';
export const CLIENT_ID = '';
export const CLIENT_SECRET = '';
