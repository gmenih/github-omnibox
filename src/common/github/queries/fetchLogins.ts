import {GraphQLClient} from 'graphql-request';
import {DEFAULT_FIRST_LOGINS} from '../constants';

interface OrganizationNode {
    login: string;
}

interface PageInfo {
    endCursor: string | null;
}

interface GithubViewerResponse {
    viewer: {
        login: string;
        name: string;
        organizations: {
            nodes: OrganizationNode[];
            pageInfo: PageInfo;
        };
    };
}

export interface GithubLogins {
    displayName?: string;
    organizations: string[];
    username: string;
}

export async function fetchLogins (
    githubClient: GraphQLClient,
    pageSize: number = DEFAULT_FIRST_LOGINS,
): Promise<GithubLogins> {
    let endCursor: string | null = null;
    let displayName: string | undefined;
    let username: string = '';
    const loginsSet: Set<string> = new Set();
    do {
        try {
            const response: GithubViewerResponse = await githubClient.request<GithubViewerResponse>(
                /* GraphQL */ `
                    query getUserInfo($pageSize: Int!, $endCursor: String) {
                        viewer {
                            login
                            name
                            organizations(after: $endCursor, first: $pageSize) {
                                nodes {
                                    login
                                }
                                pageInfo {
                                    endCursor
                                }
                            }
                        }
                    }
                `,
                {
                    endCursor,
                    pageSize,
                },
            );
            if (endCursor === null) {
                username = response.viewer.login;
                displayName = (response.viewer.name && response.viewer.name.split(' ')[0]) || undefined;
            }
            response.viewer.organizations.nodes.forEach(({login}) => {
                loginsSet.add(login);
            });
            if (response.viewer.organizations.nodes.length >= pageSize) {
                endCursor = response.viewer.organizations.pageInfo.endCursor;
            }
        } catch (error) {
            throw new Error('Could not fetch logins');
        }
    } while (endCursor !== null);
    return {
        displayName,
        organizations: [...loginsSet],
        username,
    };
}
