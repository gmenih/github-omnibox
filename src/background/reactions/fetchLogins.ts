import {reaction} from 'mobx';
import {createClient, fetchLogins} from '../../common/github/queries';
import {StorageObservable} from '../../common/storage';

export function fetchLoginsForToken (storage: StorageObservable): void {
    reaction(
        (): string | undefined => storage.token,
        async (token) => {
            if (!token) {
                return;
            }
            try {
                const loginResult = await fetchLogins(createClient(token));
                if (loginResult.username) {
                    storage.setLoginData(
                        true,
                        loginResult.username,
                        loginResult.displayName,
                        loginResult.organizations,
                    );
                }
            } catch (err) {
                storage.setLoginData(false);
            }
        },
    );
}
