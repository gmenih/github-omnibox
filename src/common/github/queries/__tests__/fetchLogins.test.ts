import {GraphQLClient} from 'graphql-request';
import {DEFAULT_FIRST_LOGINS} from '../../constants';
import {fetchLogins} from '../fetchLogins';

const stubLoginsResponse = {
    viewer: {
        login: 'test-login',
        name: 'Test Name',
        organizations: {
            nodes: [
                {
                    login: 'test-organization-1',
                },
                {
                    login: 'test-organization-2',
                },
            ],
            pageInfo: {
                endCursor: 'end-cursor',
            },
        },
    },
};

describe('Github fetchLogins', () => {
    let mockGraphQLClient: GraphQLClient;

    beforeEach(() => {
        mockGraphQLClient = {
            request: jest.fn(),
        } as any;
    });

    test('query matches snapshot and is called with proper arguments', async () => {
        let calledQuery: string = '';
        (mockGraphQLClient.request as jest.Mock).mockImplementation((query: string) => {
            calledQuery = query;
            return stubLoginsResponse;
        });

        await fetchLogins(mockGraphQLClient);

        expect(mockGraphQLClient.request).toHaveBeenCalledTimes(1);
        expect(mockGraphQLClient.request).toHaveBeenCalledWith(expect.stringContaining('query'), {
            endCursor: null,
            pageSize: DEFAULT_FIRST_LOGINS,
        });
        expect(calledQuery).toMatchSnapshot();
    });

    test('maps data to GithubLogins interface', async () => {
        (mockGraphQLClient.request as jest.Mock).mockResolvedValueOnce(stubLoginsResponse);

        const result = await fetchLogins(mockGraphQLClient);

        expect(result).toMatchSnapshot();
    });
});
