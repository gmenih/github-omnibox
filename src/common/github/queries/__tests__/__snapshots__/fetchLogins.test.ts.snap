// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Github fetchLogins maps data to GithubLogins interface 1`] = `
Object {
  "displayName": "Test",
  "organizations": Array [
    "test-organization-1",
    "test-organization-2",
  ],
  "username": "test-login",
}
`;

exports[`Github fetchLogins query matches snapshot and is called with proper arguments 1`] = `
"
                    query getUserInfo($pageSize: Int!, $endCursor: String) {
                        viewer {
                            login
                            name
                            organizations(after: $endCursor, first: $pageSize) {
                                nodes {
                                    login
                                }
                                pageInfo {
                                    endCursor
                                }
                            }
                        }
                    }
                "
`;
